File: Greedy/huffman-coding.c
// Algorithm: Huffman Coding
// Type: Greedy
// Time: O(n log n)
// Space: O(n)

#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    char ch;
    int freq;
    struct Node *left, *right;
} Node;

Node* createNode(char ch, int freq) {
    Node* node = (Node*) malloc(sizeof(Node));
    node->ch = ch;
    node->freq = freq;
    node->left = node->right = NULL;
    return node;
}

// Priority queue and heap operations would go here...

void printCodes(Node* root, char* code, int depth) {
    if (!root) return;
    if (!root->left && !root->right) {
        code[depth] = '\0';
        printf("%c: %s\n", root->ch, code);
        return;
    }
    code[depth] = '0';
    printCodes(root->left, code, depth + 1);
    code[depth] = '1';
    printCodes(root->right, code, depth + 1);
}

File: Greedy/huffman-coding.java
// Algorithm: Huffman Coding
// Type: Greedy
// Time: O(n log n)
// Space: O(n)

import java.util.*;

class Node {
    char ch;
    int freq;
    Node left, right;

    Node(char ch, int freq) {
        this.ch = ch; this.freq = freq;
    }
}

class Huffman {
    public static Node buildTree(List<Node> nodes) {
        PriorityQueue<Node> pq = new PriorityQueue<>(Comparator.comparingInt(n -> n.freq));
        pq.addAll(nodes);

        while (pq.size() > 1) {
            Node left = pq.poll();
            Node right = pq.poll();
            Node merged = new Node('\0', left.freq + right.freq);
            merged.left = left; merged.right = right;
            pq.add(merged);
        }

        return pq.poll();
    }

    public static void printCodes(Node root, String code) {
        if (root == null) return;
        if (root.left == null && root.right == null) {
            System.out.println(root.ch + ": " + code);
            return;
        }
        printCodes(root.left, code + "0");
        printCodes(root.right, code + "1");
    }
}
